### PR TITLE
Mayadoesbetter

### PR DESCRIPTION
**Pull request by mayacodes**
Updated lexical feature code :)

### CODE REVIEW

**Review by richagupta0211 — COMMENTED** [review]
Function names should be able to tell what the function does. Since a function is a process. Here you can call the function return_lexical_features so only by looking at the name of the function anyone would know what this function would do.

File: models/baseline.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+from sklearn.model_selection import train_test_split
+import matplotlib.pyplot as plt
+
+### II. Import data
+# Path to the environment variables file .env
+env_path = '/data/dssg-disinfo/.env'
+load_dotenv(env_path, override=True)
+
+
+def plot_graphs(history, string):
```
Comment (noahbenson) [inline models/baseline.py]: Would be nice if this function had a short docstring; also since the `string` parameter is really the `ylabel` it would probably be a good idea to call it `ylabel` or similar.

File: models/baseline.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    plt.plot(history.history[string])
+    plt.plot(history.history['val_'+string])
+    plt.xlabel("Epochs")
+    plt.ylabel(string)
+    plt.legend([string, 'val_'+string])
+    file_name=string+'.png'
+    plt.savefig(file_name)
+    return
```
Comment (noahbenson) [inline models/baseline.py]: A good idea is to call `plt.close()` to close the figure when you're planning on saving it instead of plotting it. (You can ignore this if you intend this function to both save and show the figure, but that kind of thing would generally go in the docstring.)

File: models/baseline.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+def LSTM_model(VOCAB_SIZE = 10000, EMBEDDING_DIM = 300, MAX_LENGTH = 681, NUM_EPOCHS = 5):
```
Comment (noahbenson) [inline models/baseline.py]: **Minor**: It's not a big deal at all, but conventionally, upper-case variable names are reserved for constants; things like `NUM_THREADS` or `GRAVITATIONAL_CONSTANT` that won't ever change. Function arguments are usually lower case like `vocab_size`, `max_length` etc.

**Big Idea**: These parameters (`VOCAB_SIZE`, `EMBEDDING_DIM`, etc.) are parameters of the **model construction**. The fact that you've setup this function to accept options that correspond to your model's flavor-parameters is good! If all of your parameters are numbers, then a simple function is all you really need to convert from a set of model flavor parameters to a reified model is the function. However, see some of my comments below about issues related to parameters that require code.

**Smaller Idea**: Another thing to think about is that using default values in these arguments might make things more confusing in the long-run. If you create a model using the code `LSTM_model()` then later decide that the default arguments should be different you now have a model that appears to have been created with a different parameter set than it previously had. At the very least, I would keep track of the exact parameters that are associated with each model explicitly and not rely on the defaults to tell you what kind of model you have.

File: models/baseline.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    MAX_LENGTH: The number of tokens that will be kept from each 
+    NUM_EPOCHS: The number of times the model will be run
+    
+    output:
+    history: The model fit output
+    """
+    # Get the paths
+    PATH = os.getenv("PATH")
+    CLEAN_DATA = os.getenv("CLEAN_DATA")
+    df = pd.read_csv(os.path.join(PATH, CLEAN_DATA))
```
Comment (noahbenson) [inline models/baseline.py]: This is a very confusing set of lines, but this is due to the fact that `PATH` is a special/reserved environment variable.

The environment variable `PATH` is usually a `:`-separated list of places to look for executable files. E.g., `"/usr/bin:/bin:/opt/bin"`.

I think you are using the `PATH` environment variable here to store your data’s path. This is a bad idea because PATH is used for telling the OS where to find executable programs like `ls` and such. If you overwrite it in your program by loading in a dotenv file containing `PATH`, you will overwrite the path and potentially prevent other python functions from working correctly. I would just change this to `DATA_PATH` or something like that.

File: models/baseline.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+
+    ### III. Splitting the data into training and testing
+    X = df['article_text'] # article_text
+    y = df.label
+    
+    training_sentences, testing_sentences, training_labels, testing_labels = train_test_split(X, y, random_state = 42)
```
Comment (noahbenson) [inline models/baseline.py]: Kseniyah and I worked through a variation of the `train_test_split` function that splits the data such that the validation dataset contained equal numbers of negative and positive articles. Let's call that function (whether it is in the repo or not) `equal_train_test_split`.

This is one choice in the **model construction** and is a good example of a problem: we essentially have a parameter of the model's flavor (should we use the `train_test_split` function or the `equal_train_test_split` function to split the data?). One way to handle this would be to add another parameter called `use_equal_split` to the function and to use `train_test_split` when `use_equal_split == False` and `equal_train_test_split` otherwise. That would be fine if there were only two possible settings.

But what if there are a lot of ways this might go? See comments on the network architecture below!

File: models/baseline.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    # making y into np arrays
+    training_labels_final = np.array(training_labels)
+    testing_labels_final = np.array(testing_labels)
+
+    vocab_size = VOCAB_SIZE 
+    embedding_dim = EMBEDDING_DIM
+    max_length = MAX_LENGTH
+    trunc_type='post'
+    oov_tok = "<OOV>"
```
Comment (noahbenson) [inline models/baseline.py]: These look to me like hidden parameters of the model. I would suggest making them explicit (i.e., put them in the function's parameter list and give them default values.)

File: models/baseline.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    testing_padded = pad_sequences(testing_sequences,maxlen=max_length)
+
+
+    ### V. Network architecture/building the model
+    model = keras.Sequential([
+        keras.layers.Embedding(vocab_size, embedding_dim, input_length=max_length),
+        keras.layers.Bidirectional(keras.layers.LSTM(64)),
+        keras.layers.Dense(64, activation='relu'),
+        keras.layers.Dense(1, activation='sigmoid')
+    ])
```
Comment (noahbenson) [inline models/baseline.py]: So this block of code is very important because it determines how the neural network is built.

It also represents something you may want to change in the future. But this presents a real problem--changes to this layer are hard to document and may take any number of forms. It would be very hard to create one parameter that succinctly described all the possible ways you might want to build a model.

There is no simple way to solve this issue. However, there are a number of strategies one can use.

One way of dealing with this, which I will call the naive way (and which I don't recommend) would be to name these builds and just add them to the function. This might look like this:
```python
def create_model(vocab_size=10000, embedding_dim=300, max_length=681, num_epochs=5, arch='basic'):
    # ... setup ...
    # Create the architecture:
    if arch == 'basic':
        model = keras.Sequential([
            keras.layers.Embedding(vocab_size, embedding_dim, input_length=max_length),
            keras.layers.Bidirectional(keras.layers.LSTM(64)),
            keras.layers.Dense(64, activation='relu'),
            keras.layers.Dense(1, activation='sigmoid')
        ])
    elif arch == '5layers':
        # construct a 5-layer network
    elif arch == '6layers':
        # construct a 6-layer network
    # ... etc. ...
```

This is bad for a lot of reasons, and one of them is that you have to maintain a big list of these in the code. A better (albeit similar) approach would be to have a registry of model architecture names and a way to register them:

```python
# The underscore at the beginning of the name of this var indicates that it is 
# private and shouldn't be modified outside of the file.
_model_arch_registry = {}
def register_model_arch(arch_name, create_fn, param_names):
    """
    register_model_arch(name, fn, param_names) registers a model architecture
    with the given name so that it can be built with the build_model() function.

    The name should be a string that identifies the model.
    The fn should be a function that accepts the parameters in the param_names and
        yields a sequential model.
    The param_names should be a list of the names of the parameters that fn requires.
    """
    # Save the model
    _model_arch_registry[arch_name] = (create_fn, param_names)

def build_model_arch(arch_name, param_dict):
    """
    Builds a model using the given architecture name and the given parameter
    dictionary.
    """
    # lookup the params and create function:
    (create_fn, params) = _model_arch_registry[arch_name]
    # The f(*[...]) syntax means that instead of being called as f([a,b,c]) the
    # functiion call is converted into f(a, b, c).
    return create_fn(*[param_dict[k] for k in params])

# Creation function for the basic model architecture:
def create_basic_model_arch(vocab_size, embedding_dim, max_length):
    "Creates and returns a basic model architecture."
    model = keras.Sequential([
        keras.layers.Embedding(vocab_size, embedding_dim, input_length=max_length),
        keras.layers.Bidirectional(keras.layers.LSTM(64)), # 64 could be a parameter?
        keras.layers.Dense(64, activation='relu'), # 64 and 'relu' could be params?
        keras.layers.Dense(1, activation='sigmoid') # 'sigmoid' could be a param?
    ])
    return model
# Register the basic model
register_model_arch("basic", create_basic_model_arch,
                    ["vocab_size, "embedding_dim", "max_length"])

# ...

def build_model(vocab_size=10000, embedding_dim=300, max_length=681, num_epochs=5, model_arch='basic'):
    """Builds a model using the passed parameters."""
    # (This next line could be implemented by using def build_model(**params) instead)
    params = {
        'vocab_size': vocab_size,
        'embedding_dim': embedding_dim,
        'max_length': max_length,
        'num_epochs': num_epochs,
        'model_arch': 'basic'
    }
    # ...build up other parts of the model...
    model = build_model_arch(params['model_arch'], params)
    # ...etc...
```
The idea behind this is that you have a single point that converts model names into model architectures, and you can add to that conversion from other places without mucking with the internals of the model building function.

Yet another way to do this would be to have classes that handled the building of different kinds of models. We can chat more about this in person.

File: models/baseline.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    model.summary()
+
+    ### VI. Put model together and run
+    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
+
+    ### VII. Fitting and running the model
+    num_epochs = NUM_EPOCHS
+    file_name = 'LSTM_model'+'_'+str(VOCAB_SIZE)+'_'+str(EMBEDDING_DIM)+'_'+str(MAX_LENGTH)+'_'+str(NUM_EPOCHS)+'.log'
+    csv_logger = CSVLogger(file_name, append=True, separator=';')
+    history=model.fit(padded, training_labels_final, epochs=num_epochs, validation_data=(testing_padded, testing_labels_final), callbacks=[csv_logger])
```
Comment (noahbenson) [inline models/baseline.py]: So here you first compile then fit and run the model. A couple things:
1. There are some implicit model parameters here (optimizer, loss).
2. Building, compiling, and fitting the model might reasonably go together, but you might think about splitting these out into separate functions. The main reason for this is that you might eventually find that you want to treat fitting the model as a parameterized step like the architecture (above comment).

File: models/baseline.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    
+    #plot_graphs(history, 'accuracy')
+    #plot_graphs(history, 'loss')
+    
+    return history
```
Comment (noahbenson) [inline models/baseline.py]: I assume that the history is sufficient to be able to reproduce the model, but you might want to return the model itself? (This depends on how tensorflow structures these things).

File: models/baseline.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
+#--------------------------------
+#       Baseline model 
+#--------------------------------
```
Comment (noahbenson) [inline models/baseline.py]: Okay, so some high-level ideas:

- First of all, this file represents the synthesis of a lot of neural-network technomancy, so it's already an accomplishment!
- Right now, what you have is a set of functions that can train a single "flavor" of model. A main problem you are facing is that you aren't sure that this flavor is exactly right--you might end up wanting to construct your test/validation sets to be slightly different sizes, or you might want to test things using a different `max_length` parameter. Different variations of these kinds of parameters are going to result in different flavors of the model that then get trained and tested.
- You want the code that constructs the different flavors of the models to be well-organized. This means that the code involved in making the models shouldn't get duplicated just because you need a new flavor, and it also means that it should be easy to convert a set of flavor-parameters into a unique model flavor and vice-versa.

Note that I'm being a little loose with the term "parameters" in the comments below; in general, when I talk about model parameters, they are in fact hyperparameters.

File: models/baseline.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    num_epochs = NUM_EPOCHS
+    file_name = 'LSTM_model'+'_'+str(VOCAB_SIZE)+'_'+str(EMBEDDING_DIM)+'_'+str(MAX_LENGTH)+'_'+str(NUM_EPOCHS)+'.log'
+    csv_logger = CSVLogger(file_name, append=True, separator=';')
+    history=model.fit(padded, training_labels_final, epochs=num_epochs, validation_data=(testing_padded, testing_labels_final), callbacks=[csv_logger])
+    
+    #plot_graphs(history, 'accuracy')
+    #plot_graphs(history, 'loss')
+    
+    return history
+
```
Comment (noahbenson) [inline models/baseline.py]: Here are a few good goals to keep in mind when organizing the various model (hyper-)parameters such as architecture and number of epochs.

1. At the end of the project, when you've tested 50 different models, each with unique hyperparameter sets, you'll want to be able to make a pandas dataframe that has one column for each hyperparameter and a column (or a few) that summarizes the model accuracy/performance. The earlier you start thinking about how you will decide what hyperparameters go in these columns the better. It's completely fine, for example, if you get halfway through the model evaluations and realize that you need to vary some new parameter that has so far always been equal to, e.g., 10. It's less fine if, halfway through, you realize that you've been varying some parameter and it hasn't been recorded. I would recommend finding out how to dump a model to disk, then writing a function that will save the model as a file and save its parameters in a separate JSON file with the same base-name or something similar (there is a json library for python that is easy to use). This way you can easily re-obtain the parameters of a model you ran and forgot about.
2. It's hard to make a chunk of code into a parameter, even though that's often what you want. In Python, classes and objects partly solve this problem, but they can't entirely solve it. You will have to be very careful around any model parameters that instruct the model-building functions to perform one block of code instead of another.
3. I'm happy to review and make suggestions on code that interfaces with this problem!

**Comment by richagupta0211** [comment]
closing this
